step/v2/step.go

```diff
@@ -0,0 +1,264 @@
+// Step is a library for building user facing pipelines.
+//
+// Step is responsible for presenting each step in a job to the user, and halting the
+// pipeline on an error.
+package step
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/briandowns/spinner"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+//go:generate go run ./generate/calls.go
+
+type pipeline struct {
+	title     string
+	callstack []string
+	failed    error
+	spinner   *spinner.Spinner
+	labels    []string
+}
+
+type pipelineKey struct{}
+
+func withPipeline(ctx context.Context, p *pipeline) context.Context {
+	return context.WithValue(ctx, pipelineKey{}, p)
+}
+
+func getPipeline(ctx context.Context) *pipeline {
+	v := ctx.Value(pipelineKey{})
+	if v == nil {
+		return nil
+	}
+	return v.(*pipeline)
+}
+
+func Pipeline(name string, steps func(context.Context)) error {
+	return PipelineCtx(context.Background(), name, steps)
+}
+
+func PipelineCtx(ctx context.Context, name string, steps func(context.Context)) (err error) {
+	if getPipeline(ctx) != nil {
+		panic("Cannot call pipeline when already in a pipeline")
+	}
+
+	p := &pipeline{
+		title: name,
+		spinner: spinner.New([]string{"|", "/", "-", "\\"},
+			time.Millisecond*250,
+			spinner.WithHiddenCursor(true),
+		),
+	}
+
+	// If we are initially silent, don't bother to create and start a spinner
+	var silent bool
+	for _, env := range getEnvs(ctx) {
+		_, silent = env.(*Silent)
+		if silent {
+			break
+		}
+	}
+
+	if silent {
+		p.spinner.Writer = io.Discard
+	}
+
+	p.setDisplay(getEnvs(ctx))
+	done := make(chan struct{})
+	go func() {
+		defer func() { close(done) }()
+		p.spinner.Start()
+		steps(withPipeline(ctx, p))
+	}()
+	<-done
+	p.setDisplay(getEnvs(ctx))
+	if p.failed == nil {
+		p.spinner.FinalMSG = fmt.Sprintf("%s--- done ---\n", p.spinner.Prefix)
+	} else {
+		p.spinner.FinalMSG = fmt.Sprintf("%s--- failed: %s ---\n",
+			p.spinner.Prefix, p.failed.Error())
+	}
+	p.spinner.Stop()
+	return p.failed
+}
+
+func mustGetPipeline(ctx context.Context, name string) *pipeline {
+	p := getPipeline(ctx)
+	if p == nil {
+		panic(`Must call "` + name + `" on a context from a step function`)
```

Comment:
Isn't it more that it needs to be called in the context of a pipeline? Also, I wonder if it's really necessary to hard fail here. You should be able to run the steps without a pipeline, you just lose the functionality the pipeline adds.